### PR TITLE
Add comprehensive method docstring tests

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This PR adds support for validating `@public`-annotated methods on top-level exported classes. The changes include:

1. Improved docstring validation for top-level exported symbols
2. Added support for validating `@public`-annotated methods on classes
3. Enhanced section header detection with a more comprehensive list of allowed section headers
4. Added utilities for working with `@public`-annotated symbols

These improvements will help ensure consistent documentation quality across both top-level symbols and public methods, making the API documentation more complete and reliable.

## How I Tested These Changes

- Added comprehensive test suite for `@public` method detection and validation
- Created test fixtures with various method types (instance, static, class, property)
- Added tests for section header detection and validation
- Verified that only `@public` methods on `@public` classes are selected for validation
- Tested edge cases like empty docstrings and various method decorators

Manual test:

Injected error and confirmed got a message

### CODE REVIEW

**Comment by schrockn** [comment]
* **#31288** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31288?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#31289](https://github.com/dagster-io/dagster/pull/31289) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31289?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#31299](https://github.com/dagster-io/dagster/pull/31299) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31299?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#31287** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31287?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31287?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#31286** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31286?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31319** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31319?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31331** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/automation/automation_tests/docstring_lint_tests/test_known_valid_symbols.py [hunk 1]
```diff
@@ -28,17 +29,63 @@
...
     "dagster.scaffold_with",  # Module resolution error
-    "dagster.sys",  # Import issue (conflicts with built-in sys)
+    # Components with YAML code snippets that should use .. code-block:: yaml instead
+    "dagster.Component",  # Contains literal YAML with "attributes:" that gets flagged as section header
+    "dagster.DefsFolderComponent",  # Contains literal YAML with "attributes:" that gets flagged as section header
+    "dagster.PythonScriptComponent",  # Contains literal YAML with "attributes:" that gets flagged as section header
+    "dagster.UvRunComponent",  # Contains literal YAML with "attributes:" that gets flagged as section header
+    # Invalid section headers that need to be fixed
+    "dagster.ExecuteInProcessResult",  # Invalid section header: "This object is returned by:" should be restructured
+    "dagster.asset_check",  # Invalid section header: "Example with a DataFrame Output:" should be "Examples:"
```
Comment (cmpadden) [inline python_modules/automation/automation_tests/docstring_lint_tests/test_known_valid_symbols.py:39]: Will be good to have a fast follow-up to resolve these.

Comment (schrockn) [reply python_modules/automation/automation_tests/docstring_lint_tests/test_known_valid_symbols.py:39]: Yup! These exist just to organize the work and make the process incremental

**Comment by schrockn** [comment]
### Merge activity

* **Jul 24, 11:05 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31287).
* **Jul 24, 11:05 PM UTC**: @schrockn merged this pull request with [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31287).
